### PR TITLE
Generate Image URL with ImageOptions

### PR DESCRIPTION
**Pull request by loudmouth**
Implement tests.

Throw errors for invalid url construction for Images API

### CODE REVIEW

File: Sources/Asset.swift [hunk 1]
```diff
@@ -12,6 +12,36 @@ import ObjectMapper
...
+
+    /// Description of the asset. Optional for compalibility with `select` operator queries.
+    public var description: String?
+
+    /// Metadata describing the file associated with the asset. Optional for compalibility with `select` operator queries.
+    public var file: FileMetadata?
+
+    public struct FileMetadata: ImmutableMappable {
+
+        let fileName: String
```
Comment (loudmouth) [inline Sources/Asset.swift]: needs documentation.

File: Sources/Asset.swift [hunk 2]
```diff
@@ -12,6 +12,40 @@ import ObjectMapper
 /// An asset represents a media file in Contentful
 public class Asset: Resource, LocalizedResource {
 
+    /// URL of the media file associated with this asset. Optional for compatibility with `select` operator queries.
+    public var urlString: String?
```
Comment (mariobodemann) [inline Sources/Asset.swift]: Hey, do you have any plans on also supporting the `upload` url field? (i.e. the one used while an asset is processing?)

File: Sources/Asset.swift [hunk 3]
```diff
@@ -25,16 +59,11 @@ public class Asset: Resource, LocalizedResource {
     public var locale: String
 
     /// The URL for the underlying media file
-    public func URL() throws -> Foundation.URL {
-        if let urlString = (fields["file"] as? [String: Any])?["url"] as? String {
-            if let URL = Foundation.URL(string: "https:\(urlString)") {
-                return URL
-            }
+    public func url() throws -> URL {
+        guard let urlString = self.urlString else { throw SDKError.invalidURL(string: "") }
```
Comment (mariobodemann) [inline Sources/Asset.swift]: That can have a nasty side effect if the asset is uploading, doesn't it?

File: Sources/Client.swift [hunk 4]
```diff
@@ -333,39 +330,20 @@ extension Client {
     /**
      Fetch the underlying media file as `Data`.
 
+     - Parameter asset: The `Asset` which contains the relevant media file.
+     - Parameter imageOptions: An optional array of options for server side manipulations.
      - Returns: Tuple of the data task and a signal for the `Data` result.
+
      */
-    public func fetchData(for asset: Asset) -> Observable<Result<Data>> {
+    public func fetchData(for asset: Asset, with imageOptions: [ImageOption] = []) -> Observable<Result<Data>> {
```
Comment (mariobodemann) [inline Sources/Client.swift]: :heart_eyes:

File: Sources/ImageOptions.swift [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+     */
+    public func url(with imageOptions: [ImageOption] = []) throws -> URL {
+
+        // Check that there are no two image options that specifiy the same query parameter.
+        // https://stackoverflow.com/a/27624476/4068264z
+        // A Set is a collection of unique elements, so constructing them will invoke the Equatable implementation
+        // and unique'ify the elements in the array.
+        let uniqueImageOptions = Array(Set<ImageOption>(imageOptions))
+        guard uniqueImageOptions.count == imageOptions.count else {
+            throw SDKError.invalidImageParameters("Cannot specify two instances of ImageOption of the same case."
```
Comment (mariobodemann) [inline Sources/ImageOptions.swift]: So, actually you can specify two options at the same time: ```http://images.contentful.com/arqlnkt58eul/2ReMHJhXoAcy4AyamgsgwQ/788a7c2210c923644b4bc20deb9a1f78/lewis-carroll-1.jpg?w=100&w=2000``` will return an image with a width of 100.

File: Sources/ImageOptions.swift [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    /// Specify that the API should return a progressive JPG.
+    /// The progressive JPEG format stores multiple passes of an image in progressively higher detail.
+    case progressive
+
+    fileprivate func urlQueryItem() throws -> URLQueryItem? {
+        switch self {
+        case .unspecified:
+            return nil
+        case .asPercent(let quality):
+            if quality > 100 {
```
Comment (mariobodemann) [inline Sources/ImageOptions.swift]: `or < 0` ?

File: Sources/ImageOptions.swift [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+    static let progressiveJPG   = "fl"
+}
+
+// Use CGColor instead of UIColor to enable cross-platform compatibility: macOS, iOS, tvOS, watchOS.
+internal extension CGColor {
+
+    // If for some reason the following code fails to create a hex string, the color black will be
+    // returned.
+    internal func hexRepresentation() -> String {
+        let hexForBlack = "ffffff"
```
Comment (mariobodemann) [inline Sources/ImageOptions.swift]: `ffffff` ➡ `0`? or `Black` ➡ `White`?

File: Sources/ImageOptions.swift [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+        }
+        return url
+    }
+}
+
+
+public enum ImageOption: Equatable, Hashable {
+
+    /// Specify the size of the image in pixels to be returned from the API. Valid ranges for width and height are between 0 and 4000.
+    case sizedTo(width: Int, height: Int)
```
Comment (mariobodemann) [inline Sources/ImageOptions.swift]: Can I specify a `width` only? Aka, I want to scale the image to 100 pixel width, and let the aspect ratio kick in to calculate the `height`?

File: Sources/Asset.swift (hunk 2 above)
Comment (loudmouth) [inline Sources/Asset.swift]: this is CDA only so far. But, i suppose in the future...yes? I'm only implementing what is listed [here](https://www.contentful.com/developers/docs/references/content-delivery-api/#/reference/assets)

File: Sources/Asset.swift (hunk 3 above)
Comment (loudmouth) [inline Sources/Asset.swift]: This is CDA only...i'm not sure i understand what you are talking about.

File: Sources/ImageOptions.swift (hunk 6 above)
Comment (loudmouth) [inline Sources/ImageOptions.swift]: it's an `UInt` so it's enforced to be `>= 0` at compile time

Comment (mariobodemann) [inline Sources/ImageOptions.swift]: ah, okay. I wasn't able to spot the type ... ;)

**Pull request merged**